torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -0,0 +1,422 @@
+  buckets_.reserve(bucket_count);
+  for (size_t bucket_index = 0; bucket_index < bucket_count; bucket_index++) {
+    Bucket bucket;
+
+    // TODO(@pietern): Validate indices.
+    // Must be non-empty, unique, and unique across buckets.
+    AT_ASSERTM(indices[bucket_index].size() > 0, "Empty bucket specified.");
+
+    // Iterate over model replicas.
+    for (size_t replica_index = 0; replica_index < replica_count;
+         replica_index++) {
+      at::TensorOptions options;
+      BucketReplica replica;
+      size_t offset = 0;
+
+      // Iterate over bucket variables.
+      for (const auto variable_index : indices[bucket_index]) {
+        AT_ASSERTM(
+            variable_index < variables_[replica_index].size(),
+            "Out of range variable index specified.");
+        const auto& variable = variables_[replica_index][variable_index];
+        if (!options.has_device()) {
+          options = options.device(variable.device());
+        } else {
+          AT_ASSERTM(
+              variable.device() == options.device(),
+              "All parameters in a bucket must be placed on the same device.");
+        }
+        if (!options.has_dtype()) {
+          options = options.dtype(variable.dtype());
+        } else {
+          AT_ASSERTM(
+              variable.dtype() == options.dtype(),
+              "All parameters in a bucket must have the same dtype.");
+        }
+        const auto length = variable.numel();
+        replica.variables.push_back(variable);
+        replica.offsets.push_back(offset);
+        replica.lengths.push_back(length);
+        offset += length;
+      }
+
+      // Allocate bucket contents tensor.
+      replica.contents = torch::autograd::make_variable_consuming(
+          at::empty({offset}, options));
+
+      // Add bucket replica to enclosing bucket.
+      bucket.replicas.push_back(std::move(replica));
+    }
+
+    // Map participating variables to this bucket.
+    // This is identical across replicas so we only need to do this once.
+    size_t intra_bucket_index = 0;
+    for (const auto variable_index : indices[bucket_index]) {
+      AT_ASSERTM(
+          variable_index < bucket_indices_.size(),
+          "Out of range variable index specified.");
+      bucket_indices_[variable_index] = BucketIndex{
+          .bucket_index = bucket_index,
+          .intra_bucket_index = intra_bucket_index++,
+      };
+    }
+
+    buckets_.push_back(std::move(bucket));
+  }
+}
+
+// Traverse the autograd graph starting at the specified output.
+// All parameters for which we have a pointer to their gradient accumulation
+// functions and don't show up in this graph can be marked as ready
+// for reduction immediately. Not doing this means we would deadlock waiting
+// on a gradient for those parameters that will never be computed.
+//
+// Rough copy of torch::autograd::Engine::compute_dependencies.
+//
+void Reducer::prepare_for_backward(const torch::autograd::Variable& output) {
```

Comment:
what if there are multiple outputs? Call it multiple times?

Reply:
Addressed this in the latest commits.